File: solve/0429_햄버거 만들기/postforty/햄버거 만들기.js
function solution(ingredient) {
  var answer = 0;
  let i = 0;

  // * 배열 비교방법간 성능은 별 차이 없었음
  // 1. every를 이용한 배열 비교
  const equals = (a, b) =>
    a.length === b.length && a.every((v, i) => v === b[i]);

  while (i < ingredient.length) {
    // 2. stringify를 이용한 배열 비교
    // if (JSON.stringify(ingredient.slice(i, i + 4)) === "[1,2,3,1]") {
    if (equals(ingredient.slice(i, i + 4), [1, 2, 3, 1])) {
      ingredient.splice(i, 4);
      answer++;
      i -= 3;
    } else {
      i++;
    }
  }
  return answer;
}

console.log(solution([2, 1, 1, 2, 3, 1, 2, 3, 1])); // 2
console.log(solution([1, 3, 2, 1, 2, 1, 3, 1, 2])); // 0
console.log(solution([1, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 1])); // 2
console.log(solution([1, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1])); // 3
console.log(solution([1, 1, 2, 3, 1, 1])); // 1

File: solve/0429_햄버거 만들기/postforty/햄버거 만들기.py
def solution(ingredient):
    answer = 0

    i = 0
    # 시간 초과 예방을 위해 순회를 최소화해야 함
    while i < len(ingredient)-3:
        if ingredient[i:i+4] == [1, 2, 3, 1]:
            del ingredient[i:i+4]
            answer += 1
            i -= 3
        else:
            i += 1
    return answer
            
print(solution([2, 1, 1, 2, 3, 1, 2, 3, 1])) # 2
print(solution([1, 3, 2, 1, 2, 1, 3, 1, 2])) # 0
print(solution([1, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 1])) # 2
print(solution([1,1,2,3,1,2,3,1,2,3,1,2,3,1])) # 3
print(solution([1,1,2,3,1,1])) # 1